hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TypeUtils.java

```diff
@@ -0,0 +1,328 @@
+        private final int typeKind;
+        private final RelationsStorage relationsStorage;
+
+        FieldEnricher(int typeKind, RelationsStorage relationsStorage) {
+            this.typeKind = typeKind;
+            this.relationsStorage = relationsStorage;
+        }
+
+        void enrich(MappingField field, Map<String, String> mappingOptions, boolean isKey) {
+            String typeName = field.type().getObjectTypeName();
+            field.setType(createFieldType(
+                    field.name().equals(isKey ? QueryPath.KEY : QueryPath.VALUE)
+                            ? () -> getSchemaId(mappingOptions, isKey)
+                            : () -> getFieldSchemaId(getSchema(getSchemaId(mappingOptions, isKey)),
+                                    field.plainExternalName(), typeName),
+                    typeName, new HashMap<>()));
+        }
+
+        protected QueryDataType createFieldType(Supplier<ID> schemaIdSupplier, String typeName,
+                                                Map<String, QueryDataType> seen) {
+            QueryDataType convertedType = seen.get(typeName);
+            if (convertedType != null) {
+                return convertedType;
+            }
+
+            Type type = relationsStorage.getType(typeName);
+            if (type == null) {
+                throw QueryException.error("Encountered type '" + typeName + "', which doesn't exist");
+            }
+
+            ID schemaId = getSchemaId(type.options(), null);
+            if (schemaId == null) {
+                schemaId = schemaIdSupplier.get();
+            }
+
+            S schema = getSchema(schemaId);
+            if (type.getFields().isEmpty()) {
+                type.setFields(resolveFields(schema));
+                relationsStorage.put(typeName, type);
+            }
+
+            convertedType = new QueryDataType(typeName, typeKind);
+            convertedType.setObjectTypeMetadata(getTypeMetadata(schemaId));
+            seen.put(typeName, convertedType);
+
+            for (TypeField field : type.getFields()) {
+                QueryDataType fieldType = field.getQueryDataType();
+                String fieldTypeName = fieldType.getObjectTypeName();
+
+                if (fieldType.isCustomType()) {
+                    fieldType = createFieldType(() -> getFieldSchemaId(schema, field.getName(), fieldTypeName),
+                            fieldTypeName, seen);
+                }
+                convertedType.getObjectFields().add(new QueryDataTypeField(field.getName(), fieldType));
+            }
+
+            return convertedType;
+        }
+
+        protected abstract String getTypeMetadata(ID schemaId);
+        protected abstract S getSchema(ID schemaId);
+        protected abstract List<TypeField> resolveFields(S schema);
+        protected abstract ID getFieldSchemaId(S schema, String fieldName, String fieldTypeName);
+        protected abstract ID getSchemaId(Map<String, String> options, Boolean isKey);
```

Comment:
in what case `isKey` will be null - what does it mean?

Reply:
`isKey` should be null if you pass type options and non-null for mapping options.

Reply:
if we want to keep boolean, maybe it would make less error-prone api if we had a method with `boolean isKey` and second method without this parameter at all?

Reply:
I think it is convenient that a single method is responsible for all `KEY`/`VALUE`/`TYPE` schema ID extraction and validation. After switching to enum, it will be self-explaining.

Reply:
70681845